new module: apache2_mod_proxy

oboukili opened the pull request:
##### ISSUE TYPE
```
 - New Module Pull Request
```
##### COMPONENT NAME
```
httpd_proxy
```

##### ANSIBLE VERSION
```
ansible 2.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides

```
##### SUMMARY
```
Set or get members' attributes of an Apache httpd 2.4 mod_proxy balancer pool, using HTTP POST and GET requests.
```

(edited out the command output)

oboukili commented:
Hi,
This module basically allows people using mod_proxy enabled apache httpd 2.4+ servers as reverse proxies or http loadbalancers, to integrate a "rolling-updates" workflow using Ansible (drain member -> check for absence of connections -> disable -> update member -> hot_standby/enable; rince and repeat with other members).

Tested with python 2.6 and 2.7.
Cheers,

alikins commented:
Should this match the naming scheme for web_infrastructure/apache2_module.py. Maybe something like 'apache2_mod_proxy' ?

alikins commented on network/httpd_proxy.py:
```diff
@@ -0,0 +1,344 @@
+    HAS_BEAUTIFULSOUP = True
+
+try:
+    import requests
```
Would it be possible to let this use ansibles url handling instead of requests? That would avoid a req on the managed nodes.

alikins commented on network/httpd_proxy.py:
```diff
@@ -0,0 +1,344 @@
+    else:
+        return None
+
+class balancer_member(object):
```
CamelCase is preferred for class names. ie, 'BalancerMember'

alikins commented on network/httpd_proxy.py:
```diff
@@ -0,0 +1,344 @@
+#!/usr/bin/env python
```
Should be '#!/usr/bin/python' as per http://docs.ansible.com/ansible/developing_modules.html#module-checklist

alikins commented:
Any way to support a 'state' arg?, ala the 'Arguments' section from http://docs.ansible.com/ansible/developing_modules.html#module-checklist:
"Try and use state present and state absent like other modules"

alikins commented on network/httpd_proxy.py:
```diff
@@ -0,0 +1,344 @@
+- httpd_proxy: balancer_vhost="{{ vhost_ip }}" member_host="{{ member.ip }}" member_set_status=true draining_mode=true delegate_to=myloadbalancernode
+- wait_for: host="{{ member.ip }}" port={{ member.port}} state=drained delegate_to=myloadbalancernode
+- httpd_proxy: balancer_vhost="{{ vhost_ip }}" member_host="{{ member.ip }}" member_set_status=true disabled=true delegate_to=myloadbalancernode
+
```
This next blurb could be reformatted and used for the 'RETURN' documentation, per http://docs.ansible.com/ansible/developing_modules.html#module-documenting

sivel commented:
Some notes from the automated linting:

```
============================================================================
./network/httpd_proxy.py
============================================================================
TRACE:
    while parsing a block collection
      in "<string>", line 11, column 7:
              - [ipv4|ipv6|fqdn][:port] of the ... 
              ^
    expected <block end>, but found '['
      in "httpd_proxy.DOCUMENTATION", line 13, column 25:
              - [ipv4|ipv6|fqdn][:port] of the Apache httpd 2.4  ... 
                                ^
ERROR: DOCUMENTATION is not valid YAML. Line 13 column 25
ERROR: No RETURN documentation provided
ERROR: requests import found, should use ansible.module_utils.urls instead
ERROR: GPLv3 license header not found
ERROR: Interpreter line is not "#!/usr/bin/python"
```

sivel commented:
needs_revision

gregdek commented:
Thanks @oboukili for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

oboukili commented:
Alright, thanks for the feedback, i'll work on it.

oboukili commented:
ready_for_review

oboukili commented:
ready_for_review

gregdek commented:
Thanks @oboukili for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

oboukili commented:
Another PR on devel branch seems(?) to make this PR testing build to fail, could someone give me a much appreciated pointer?
Thanks,

oboukili commented:
nevermind, reading issues, working on it.

oboukili commented:
ready_for_review

ricsipontaz commented:
@oboukili 
Hi,
I'm looking for a loadbalancer-manager solution for Ansible and I've just found this. But when I try to use it, I get the following error:
http://pastebin.ubuntu.com/20432467/

Here's the a relevant part of my playbook:
http://pastebin.ubuntu.com/20432642/

Do you have any idea/advice? Thanks

oboukili commented:
Woopsy, thanks for the thumbs up, fixed.

oboukili commented:
The number of GET requests should be at least halved now, depending on the number of nodes you have in one vhost.

gregdek commented:
Thanks @oboukili for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

oboukili commented:
ready_for_review

oboukili commented:
shippable test results is due to python2.4 being not available on the container running tests, can't really fix that by myself :(

gregdek commented:
Thanks @oboukili for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

ricsipontaz commented:
@oboukili Thanks for the fix, but I have an other problem now:
http://paste.ubuntu.com/20971961/

Could you take a look at it?

oboukili commented:
I can't reproduce this exception.
Can you paste the output of 
`/usr/bin/python -c "exec(\"import re,sys\\nprint re.__version__\\nprint sys.version_info\")"`

oboukili commented:
Lastly, could you replace the contents of lines 189-192 with the following snippet:

`EXPRESSION = "(b=([a-z_A-Z-0-9]*)&w=(https?|ajps?)://(([0-9]{1,3}\\.?){4}):?([0-9]*)?(/.*)?&nonce=(([a-f]*[0-9]*-*)*))"`
`APACHE_VERSION_EXPRESSION = "Server Version: Apache/([0-9\\.?]*) (([aA-zZ]*))"`

I doubled escaped all backslashes.

ricsipontaz commented:
@oboukili 

```
[root@localhost ansible]#  /usr/bin/python -c "exec(\"import re,sys\\nprint re.__version__\\nprint sys.version_info\")"
2.2.1
sys.version_info(major=2, minor=7, micro=5, releaselevel='final', serial=0)
```

The same error message with the changes (double escaped blackslashes):

```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: sre_constants.error: nothing to repeat
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_9xZXHw/ansible_module_apache2_mod_proxy.py\", line 423, in <module>\n    main()\n  File \"/tmp/ansible_9xZXHw/ansible_module_apache2_mod_proxy.py\", line 367, in main\n    for member in mybalancer.members:\n  File \"/tmp/ansible_9xZXHw/ansible_module_apache2_mod_proxy.py\", line 328, in get_balancer_members\n    yield BalancerMember(str(self.base_url + balancer_member_suffix), str(self.url), self.module)\n  File \"/tmp/ansible_9xZXHw/ansible_module_apache2_mod_proxy.py\", line 218, in __init__\n    self.host = str(regexp_extraction(management_url, EXPRESSION, 4))\n  File \"/tmp/ansible_9xZXHw/ansible_module_apache2_mod_proxy.py\", line 196, in regexp_extraction\n    regexp_search = re.search(str(_regexp), str(string))\n  File \"/usr/lib64/python2.7/re.py\", line 142, in search\n    return _compile(pattern, flags).search(string)\n  File \"/usr/lib64/python2.7/re.py\", line 242, in _compile\n    raise error, v # invalid expression\nsre_constants.error: nothing to repeat\n", "module_stdout": "", "msg": "MODULE FAILURE", "parsed": false}
```

oboukili commented:
Alright, I changed the regexps, seems it was a long standing bug with some versions of Python when there 's an unescaped dash - and an unescaped equal sign = in the same regexp bracket. (read some mailing lists from 2000(!) here and there about that).
Added a few more things while I was at it. Let me know if that finally works for you.

gregdek commented:
Thanks @oboukili for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

ricsipontaz commented:
@oboukili Thanks, you solved this issue, but unfortunately I have another one:

```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: TypeError: int() argument must be a string or a number, not 'NoneType'
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_0z2gM5/ansible_module_apache2_mod_proxy.py\", line 430, in <module>\n    main()\n  File \"/tmp/ansible_0z2gM5/ansible_module_apache2_mod_proxy.py\", line 372, in main\n    for member in mybalancer.members:\n  File \"/tmp/ansible_0z2gM5/ansible_module_apache2_mod_proxy.py\", line 333, in get_balancer_members\n    yield BalancerMember(str(self.base_url + balancer_member_suffix), str(self.url), self.module)\n  File \"/tmp/ansible_0z2gM5/ansible_module_apache2_mod_proxy.py\", line 225, in __init__\n    self.port = int(regexp_extraction(management_url, EXPRESSION, 5))\nTypeError: int() argument must be a string or a number, not 'NoneType'\n", "module_stdout": "", "msg": "MODULE FAILURE", "parsed": false}
```

oboukili commented:
Alright, seems there's no port member specified in the member url, i'll remove the individual type casts for every member attribute, and will let the helper regexp_extraction function always return a string.
This is a case I haven't encountered with our vhosts definitions yet. I'll work on it later today.

This way, users will have to typecast manually desired attributes in an Ansible task (e.g. member.port|int).

Could you paste your apache vhost definition by chance? It would be much appreciated to run some tests before commiting.

ricsipontaz commented:
@oboukili It is just a simple tesf config with only one host:

```
<VirtualHost *:80>
        ProxyRequests off
        ServerName 192.168.1.141
        <Proxy balancer://testcluster>
                BalancerMember http://192.168.1.141 route=test1 min=2 smax=5 max=20 timeout=300 retry=60
                Require all granted
                ProxySet lbmethod=byrequests

                order allow,deny
                allow from all
        </Proxy>

        <Location /balancer-manager>
                SetHandler balancer-manager
        </Location>

        ProxyPass /balancer-manager !
        ProxyPass / balancer://testcluster/ stickysession=JSESSIONID|jsessionid scolonpathdelim=On
        ProxyPassReverse  / http://192.168.1.141
</VirtualHost>
```

oboukili commented:
ready_for_review

gregdek commented:
Thanks @oboukili for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

ricsipontaz commented:
Thanks, it's working now :)

```
TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": {
        "changed": false, 
        "members": [
            {
                "attributes": {
                    "Busy": "0", 
                    "Elected": "3", 
                    "Factor": "1", 
                    "From": "  0 ", 
                    "Load": "0", 
                    "Route": "test1", 
                    "RouteRedir": null, 
                    "Set": "0", 
                    "Status": "Init Err ", 
                    "To": "  0 ", 
                    "Worker URL": null
                }, 
                "balancer_url": "http://192.168.0.100/balancer-manager/", 
                "host": "192.168.1.141", 
                "management_url": "http://192.168.0.100/balancer-manager/?b=testcluster&w=http://192.168.1.141&nonce=aea0fbb2-85b5-48f9-8b11-f471b859c6e0", 
                "path": null, 
                "port": null, 
                "protocol": "http", 
                "status": {
                    "disabled": false, 
                    "drained": false, 
                    "hot_standby": false, 
                    "ignore_errors": false
                }
            }
        ]
    }
}
```

oboukili commented:
Great news! :)
Now to get this PR shipped...!

alexandrezfs commented:
Hi,
I tested this module on my side and this is working well from what I needed to do with my 3 nodes LB

Quick LB check
`ansible/hacking/test-module -m apache2_mod_proxy.py -a "balancer_vhost=xx.xx.xx.xx"`

```
***********************************
PARSED OUTPUT
{
    "changed": false,
    "invocation": {
        "module_args": {
            "balancer_url_suffix": "/balancer-manager/",
            "balancer_vhost": "xx.xx.xx.xx",
            "member_host": null,
            "state": null,
            "tls": false,
            "validate_certs": true
        }
    },
    "members": [
        {
            "attributes": {
                "Busy": "251",
                "Elected": "562",
                "Factor": "1",
                "From": "2.4K",
                "Load": "0",
                "Route": "member1",
                "RouteRedir": null,
                "Set": "0",
                "Status": "Init Ok ",
                "To": "2.6M",
                "Worker URL": "http://xx.xx.xx.xx"
            },
            "balancer_url": "http://xx.xx.xx.xx/balancer-manager/",
            "host": "xx.xx.xx.xx",
            "management_url": "http://xx.xx.xx.xx/balancer-manager/?b=hello-balancer&w=http://xx.xx.xx.xx&nonce=8b8e4a79-0d8f-4ffe-95d1-bcccc0b244dc",
            "path": null,
            "port": null,
            "protocol": "http",
            "status": {
                "disabled": false,
                "drained": false,
                "hot_standby": false,
                "ignore_errors": false
            }
        },
        {
            "attributes": {
                "Busy": "0",
                "Elected": "0",
                "Factor": "2",
                "From": "  0 ",
                "Load": "0",
                "Route": "member2",
                "RouteRedir": null,
                "Set": "1",
                "Status": "Init Ok ",
                "To": "  0 ",
                "Worker URL": "http://yy.yy.yy.yy"
            },
            "balancer_url": "http://xx.xx.xx.xx/balancer-manager/",
            "host": "yy.yy.yy.yy",
            "management_url": "http://xx.xx.xx.xx/balancer-manager/?b=hello-balancer&w=http://yy.yy.yy.yy&nonce=8b8e4a79-0d8f-4ffe-95d1-bcccc0b244dc",
            "path": null,
            "port": null,
            "protocol": "http",
            "status": {
                "disabled": false,
                "drained": false,
                "hot_standby": false,
                "ignore_errors": false
            }
        },
        {
            "attributes": {
                "Busy": "0",
                "Elected": "94",
                "Factor": "1",
                "From": "826 ",
                "Load": "0",
                "Route": "member3",
                "RouteRedir": null,
                "Set": "0",
                "Status": "Init Dis ",
                "To": "562K",
                "Worker URL": "http://127.0.0.1"
            },
            "balancer_url": "http://xx.xx.xx.xx/balancer-manager/",
            "host": "127.0.0.1",
            "management_url": "http://xx.xx.xx.xx/balancer-manager/?b=hello-balancer&w=http://127.0.0.1&nonce=8b8e4a79-0d8f-4ffe-95d1-bcccc0b244dc",
            "path": null,
            "port": null,
            "protocol": "http",
            "status": {
                "disabled": true,
                "drained": false,
                "hot_standby": false,
                "ignore_errors": false
            }
        }
    ]
}
```

Also tested to put a LB member on an absent state

`ansible/hacking/test-module -m apache2_mod_proxy.py -a "balancer_vhost=xx.xx.xx.xx member_host=127.0.0.1 state=absent"
`

```
***********************************
PARSED OUTPUT
{
    "changed": true,
    "invocation": {
        "module_args": {
            "balancer_url_suffix": "/balancer-manager/",
            "balancer_vhost": "xx.xx.xx.xx",
            "member_host": "127.0.0.1",
            "state": "absent",
            "tls": false,
            "validate_certs": true
        }
    },
    "member": {
        "attributes": {
            "Busy": "83",
            "Elected": "94",
            "Factor": "1",
            "From": "  0 ",
            "Load": "0",
            "Route": "member3",
            "RouteRedir": null,
            "Set": "0",
            "Status": "Init Dis ",
            "To": "562K",
            "Worker URL": "http://127.0.0.1"
        },
        "balancer_url": "http://xx.xx.xx.xx/balancer-manager/",
        "host": "127.0.0.1",
        "management_url": "http://xx.xx.xx.xx/balancer-manager/?b=hello-balancer&w=http://127.0.0.1&nonce=8b8e4a79-0d8f-4ffe-95d1-bcccc0b244dc",
        "path": null,
        "port": null,
        "protocol": "http",
        "status": {
            "disabled": true,
            "drained": false,
            "hot_standby": false,
            "ignore_errors": false
        }
    }
}
```

Would be definitely useful, thanks for this PR @oboukili

oboukili commented:
@alexzhxin @ricsipontaz thanks!
Would you be so kind as to tag this PR as "worksforme"? And even kinder as to review the code then tag it as "passes_guidelines" ?
Excerpt from the REVIEWERS.md ( https://github.com/ansible/ansible-modules-extras/blob/devel/REVIEWERS.md )

(edit: seems it's 'shipit' now instead of 'worksforme'...)

ricsipontaz commented:
As I see new modules still need a "worksforme" tag.

ricsipontaz commented:
worksforme

alexandrezfs commented:
worksforme

gundalow commented:
Hi,
Can you please move this file into `web_infrastructure`

needs_revision

gregdek commented:
Thanks @oboukili for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

oboukili commented:
ready_for_review

oboukili commented:
Hi,
Just rebased and moved to appropriate folder, checks are ok now.

oboukili commented:
ready_for_review

gregdek commented:
Thanks @oboukili for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

gundalow commented on web_infrastructure/apache2_mod_proxy.py:
```diff
@@ -0,0 +1,430 @@
+      - Desired state of the member host.
+    required: false
+    choices: ["present", "absent", "enabled", "disabled", "drained", "hot_standby", "ignore_errors"]
+    notes:
```
I'm not sure if you can put `notes:` under a parameter. I think it can only go at the top level.

Can you please add this to `description:`

gundalow commented:
needs_revision. Only a minor thing.

gregdek commented:
Thanks @oboukili for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

gundalow commented:
Hi,
If you put `ready_for_review` in a comment that will instruct Ansibull-bot that this is ready for review

oboukili commented:
Yes, I was waiting for the builds to pass.
ready_for_review

gundalow commented:
shipit

gregdek commented:
Thanks again to @oboukili for this PR, and thanks @ansible/core for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

gundalow commented:
Thanks!
